Fix #13876 (Document and rewrite cstyleCast) (#7539)
danmar/cppcheck: man/checkers/cstyleCast.md

# cstyleCast

**Message**: C-style pointer casting<br/>
**Category**: Type Safety<br/>
**Severity**: Style<br/>
**Language**: C++, not applicable for C code

## Motivation

C style casts can be dangerous in many ways:
 * there can be unintentional loss of precision
 * there can be unintentional loss of sign
 * there can be unintentional loss of constness
 * there can be invalid type conversions

## Philosophy

This checker warns about old style C casts that looks dangerous.

This checker is not about readability. It is about safety.

We only want to warn if there are possible safety issues and a C++ cast can provide better safety.

## Other tools / checkers

Related checkers:
 * Complements: Misra C 11.8 is violated if there is "loss of constness"
 * More noisy: Misra C++ 8.2.2, cover all C style casts.
 * More noisy: C style cast compiler warnings (i.e. clang reports Wold-style-cast)
 * Complements: Loss of constness compiler warnings.

> ## Example

Reviewer: Having examples in the documentation obviously helps but it provides incomplete code which might be misleading.

We also have the `examples` which would provide way more details and an actual sample file to analyze. Maybe we should merge that with the documentation.

Author: > Having examples in the documentation obviously helps but it provides incomplete code which might be misleading.

I made the code complete

> We also have the examples which would provide way more details and an actual sample file to analyze. Maybe we should merge that with the documentation.

I do like to have the code directly in the documentation. If we merge it somehow I am not against it basically. But personally I would drop the examples folder and just provide the documentation.

Reviewer: Having the `examples` separate helps making sure that the code is valid and is always detected. Also people use this to test their setups. We could inject the examples via a script.

Author: > We could inject the examples via a script.

ok

Reviewer: > Having the `examples` separate helps making sure that the code is valid and is always detected. Also people use this to test their setups. We could inject the examples via a script.

This is not just personal preference (i.e. avoid duplicated/manual work) but based on actual feedback I have seen with Cppcheck and clang-tidy (they have/had several broken code examples in documentation) - mostly from first-time users.

I am not blocking this PR but it is something we should consider moving forward.

We could also leverage the effort of documenting/adding examples for the checks to fix the holes in the unit testing - see https://trac.cppcheck.net/ticket/12232.

### Compliant C style cast

The goal is to not warn about "safe" and/or "intentional" casts.

For example:

> ```

Reviewer: code blocks should use `cpp`

int *p = (int*)0;
```
A static_cast can be used however technically it's the same as the C style cast.

Other example:
```
uint8_t x = (uint8_t)0x1234;
```
A static_cast can be used however technically it's the same as the C style cast.

### Non compliant C style cast

A cast from a base class pointer to a derived class pointer is potentially unsafe:
```
Derived *p = (Derived*)base;
```
A `dynamic_cast` is safer than the C-style cast here.

## How to fix

You can use `dynamic_cast` or `static_cast` to fix warnings.

Before:
```
Derived *p = (Derived*)base;
```

After:
```
Derived *p = dynamic_cast<Derived*>(base);
```

## Enable

> --enable=style

Reviewer: This seems redundant given the severity is mentioned above.

# cstyleCast

**Message**: C-style pointer casting<br/>
**Category**: Type Safety<br/>
**Severity**: Style<br/>
**Language**: C++, not applicable for C code

## Motivation

C style casts can be dangerous in many ways:
 * loss of precision
 * loss of sign
 * loss of constness
 * invalid type conversion

## Philosophy

> This checker warns about old style C casts that looks dangerous.

Reviewer 2: ```suggestion
This checker warns about old style C casts that can be dangerous.
```

This checker is not about readability. It is about safety.

This checker focus on invalid type conversions:
If there is loss of sign or precision in the cast; it's hard to know for Cppcheck if that is intentional.
If there is loss of constness; there are other checkers available and these apply to C code also.
Therefore we only warn about pointer/reference casts that could be invalid type conversions.

## Other tools / checkers

> Extended checking; if you want more warnings that warn about all c style casts:

Reviewer 2: ```suggestion
Extended checking; if you want more warnings that warn about all C style casts:
```

 * Misra C++ 8.2.2, cover all C style casts.

>  * C style cast compiler warnings (i.e. clang reports Wold-style-cast)

Reviewer 2: ```suggestion
 * C style cast compiler warnings (i.e. clang reports -Wold-style-cast)
```

Complementing; checking for loss of constness etc:
 * Misra C 11.8

>  * Loss of constness compiler warnings (i.e. clang reports Wcast-qual).

Reviewer 2: ```suggestion
 * Loss of constness compiler warnings (i.e. clang reports -Wcast-qual).
```
